### ISSUE TITLE
TMM Implement party system and team matchmaking

### ISSUE THREAD

**Issue by Geosearchef** [FAForever/server#422]

**Comment by Sheeo** [comment]
Suggested implementation and additional sideproject while we're at it (RFC):

Currently players queue up at random intervals and match with decreasing game quality over time (I cannot recall the threshold but you will get theoretically worse balanced games as time goes on in the queue). This doesn't work well when you don't have many people queuing up at once.

Instead I suggest we make queues "pop" at global times, say every 10-15 minutes for each queue (1v1, 2v2, 3v3 etc.).

This way--ideally anyway--the set of players to match is bigger and the chance of getting a balanced match is higher, at the cost of having to schedule your playing start time at queue pop interval, which might even be considered a benefit since you know exactly when your game will launch if you get a match.

Exact pop times should be variable with online population.

Pop times should be made very visible in the client so it's clear when you should queue and how much time is left to make it in this round
Ideally they're staggered as well, so there's a shorter timeframe between a 1v1 pop and a 2v2 pop than the 15 minutes to next 1v1 pop

I think the matchmaking system wholly overlaps though
There's no reason the same queue/matchmaking system cannot be used for 1v1/ladder and NvN[vN...] (edited) 

Rather, it can be described with a generic class configured with:
- Map pool
- Team configuration
- Game launch options
- Rating scheme
- (Queue pop timings) (edited) 
So, with what you've added now there's the ability to create and destroy parties. The matchmaking queue then needs to be generalized with the above support, and to deal with matching `List[Union[PlayerParty,Player]]` rather than just matching `List[Player]`
(Or to simplify things you enforce party size == queue team size in order to be able to queue your party)

**Comment by Geosearchef** [comment]
@ sheeo, that's a really good idea and we should be option for that, some other notes from various debates/private discussions:

- group up with friends (party)
- join queue
- matching at dynamic intervals
- 20-30 secs before matching, everybody in a party has to ready up
- all ready parties are considered for matching

**Comment by Askaholic** [comment]
I was thinking about making the server send the client more info during matchmaking to let the client know when a match is found. I was thinking it would look something like this:
```
{
    "command": "match_found",
    "queue": "ladder1v1",
    "opponents": [<opponent_player_id>],
    "factions": ["cybran"]
}
```

That way the client could let you know who you get matched up against, and maybe even show you a bit of their profile (finally a reason for those profile images). You could see their rank, and games played etc and what faction they're playing.

Then of course I also want to send messages when the server detects that one of the players has failed to host/join so that the other clients can know to stop waiting. For that message I was thinking of adding the offending playerid as well so that you can know if it was someone in your party for example, and you can kick them if it turns out to be a recurring problem.

```
{
    "command": "game_launch_timeout",
    "player": <offending_player_id>
}
```

I guess one potential issue I see is that if people see ahead of time that they are playing a difficult opponent, they might just cancel the automatch to prevent rating loss. We would need to account for this by either counting that somehow, or maybe just show the player what factions they're playing against and not what the ratings are.

**Comment by Askaholic** [comment]
Need some way of limiting who can send you invites to avoid streamers being spammed with invites. Maybe limit invites to friends only, or maybe a way to disable them altogether (so that you would be the one sending the invites)

**Comment by Askaholic** [comment]
Closed by #423
